File: nls-util/src/nls-util.cpp
/**
 * @file nls-util.cpp
 * @author ChenPi11
 * @brief C++ Plus NLS util.
 * @version 0.0.1
 * @date 2023-7-24
 * @copyright Copyright (C) 2023 The C++ Plus Project
 */
/* Copyright (C) 2023 The C++ Plus Project.
   This file is part of the build-aux library.

   The build-aux library is free software; you can redistribute it
   and/or modify it under the terms of the GNU General Public
   License as published by the Free Software Foundation; either version 3
   of the License, or (at your option) any later version.

   The build-aux library is distributed in the hope that it will be
   useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public
   License along with the build-aux library; see the file COPYING.
   If not, see <https://www.gnu.org/licenses/>.  */

#if _MSC_VER >= 1600 
#pragma execution_character_set("utf-8") 
#endif

#include "langmap.hpp"

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>

/**
 * @brief NLS Util main domain's name.
*/
constexpr auto MAIN_DOMAIN = "main";

/**
 * @brief Read all data in a file to a string.
 * @param file_path Input file path.
 * @return Data in the file.
 * @throw Call `abort()` when file open error.
*/
std::string readall(const std::string& file_path)
{
    std::ifstream file(file_path);
    std::stringstream buffer;
    
    if (file)
    {
        buffer << file.rdbuf();
    }
    else
    {
        perror(file_path.c_str());
        abort();
    }
    
    std::string result = buffer.str();
    file.close();
    return result;
}

/**
 * @brief Truncate output file and write data in it.
 * @param file_path Output file path.
 * @param data The data you want to write in it.
 * @throw Call `abort()` when open file error.
*/
void writeall(const std::string& file_path, const std::string& data)
{
    std::ofstream output_file(file_path, std::ios::trunc);
    if(output_file)
    {
        output_file.write(data.data(), data.size());
        output_file.flush();
        output_file.close();
    }
    else
    {
        perror(file_path.c_str());
        abort();
    }
}

/**
 * @brief Replace A to B in a string.
 * @param src Source string.
 * @param from Substring A.
 * @param to Substring B.
*/
void replace(std::string& src, const std::string& from, const std::string& to)
{
    std::string::size_type pos = 0;
    while((pos = src.find(from, pos)) != std::string::npos)
    {
        src.replace(pos, from.size(), to);
        pos += to.size();
    }
}

/**
 * @brief Input file path.
*/
std::string input_file_path;

/**
 * @brief Output file path.
*/
std::string output_file_path;

/**
 * @brief Language map file path.
*/
std::string lang_map_file_path;

// Usage: nls-util input-file output-file language-map-file
int main(int argc, char* argv[])
{
    if(argc < 4)
    {
        std::cerr << "Usage: " << argv[0] << " input-file output-file language-map-file\n";
        return 1;
    }

    input_file_path = argv[1];
    output_file_path = argv[2];
    lang_map_file_path = argv[3];

    try
    {
        load_file(lang_map_file_path, MAIN_DOMAIN);
        auto msgs = list_messages(MAIN_DOMAIN);
        
        std::string data;
        data = readall(input_file_path);
        size_t replaced_count = 0;
        for(auto& msg : msgs)
        {
            replace(data, msg, get_message(msg, MAIN_DOMAIN, msg));
            replaced_count++;
        }
        writeall(output_file_path, data);
        std::cerr << output_file_path << ": Successfully replaced " << replaced_count << " positions.\n";
        return 0;
    }
    catch(const std::exception& e)
    {
        std::cerr << e.what() << "\n";
        return 1;
    }
}

File: nls_util/nls_util.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# -*- mode: python -*-
# vi: set ft=python :


"""
C++ Plus NLS Util

Copyright (C) 2023 The C++ Plus Project.
This file is part of the build-aux library.

The build-aux library is free software; you can redistribute it
and/or modify it under the terms of the GNU General Public
License as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

The build-aux library is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public
License along with the build-aux library; see the file COPYING.
If not, see <https://www.gnu.org/licenses/>.
"""

__all__ = ["readall", "writeall", "load_file", "list_messages", "get_message", "main"]
__author__ = "LYF511, ChenPi11, The C++ Plus Project"
__license__ = "GPL-3.0-or-later"
__version__ = "0.0.1"
__maintainer__ = "ChenPi11"
__doc__ = \
"""
C++ Plus NLS Util
Read, get and replace messages in a cppp language map file.
"""

import sys

def readall(file_path):
    """Read all of the bytes data in a file and return it.

    Args:
        file_path (str): The file's path.

    Returns:
        bytes: Data in the file.
    """

    with open(file_path, "rb") as file:
        file.seek(0)
        return file.read()


def writeall(file_path, data):
    """Write all of the data into a file.

    Args:
        file_path (str): The file's path.
        data (bytes): Bytes data.
    """

    with open(file_path, "wb") as file:
        file.seek(0)
        file.write(data)
        file.flush()


def load_file(path, domain):
    """Load a language map file and return the language map.

    Args:
        path (str): Language map file.
        domain (str): The domain you want to save.

    Returns:
        dict[dict[str, bytes]]: Result language map.
    """

    line_count = 0
    loaded_count = 0
    status = 0  # 0: close, 1: key, 2: content
    key_buffer = b""
    content_buffer = b""

    # Result map.
    messages = {}

    with open(path, "rb") as file:
        lines = file.readlines()
        for line in lines:
            line_count += 1
            line = line.strip()

            if line or status:
                if (
                    line.startswith(b"#") and not status
                ):  # Skip notes when it is not in text area.
                    continue
                elif line == b"'''":
                    if status < 2:
                        status += 1
                    else:
                        status = 0
                        messages.setdefault(domain, {})[key_buffer] = content_buffer
                        key_buffer = b""
                        content_buffer = b""
                        loaded_count += 1
                elif status:  # Contents
                    if status == 1:  # Key
                        key_buffer += (b"\n" + line) if key_buffer else line
                    else:  # Content
                        content_buffer += (b"\n" + line) if content_buffer else line
                else:  # Syntax error?
                    sys.stderr.write(
                        "WARNING: " + path + ":" + str(line_count) +
                            " : Syntax error in this line.\n"
                        )

    if status:
        sys.stderr.write("WARNING: A token never closed.\n")

    sys.stderr.write(path + ": Loaded " + str(loaded_count) + " messages.\n")
    return messages

def list_messages(messages, domain):
    """List messages in a language map.

    Args:
        messages (dict): The language map.
        domain (str): Domain.

    Returns:
        list[bytes]: Result list.
    """

    if domain not in messages:
        return []

    return list(messages[domain].keys())


def get_message(messages, key, domain, default_value=""):
    """Get a message in the language map.

    Args:
        messages (dict): The language map.
        key (str): Messgae key.
        domain (str): Domain.
        default_value (str, optional): Default value if get failed. Defaults to "".

    Raises:
        ValueError: If domain or key invalid.

    Returns:
        byte: Message data.
    """

    if domain not in messages:
        raise ValueError("Invalid domain '" + domain + "'")

    domain_unit = messages[domain]
    if key not in domain_unit:
        if not default_value:
            raise ValueError("Invalid key '" + key + "'")
        return default_value
    return domain_unit[key]


# Main domain, we only used the domain named 'main' for replace files.
MAIN_DOMAIN = "main"


def main(argv):
    """C++ Plus NLS Util main entry.

    Args:
        argv (list): Arguments.

    Returns:
        int: Return value.
    """

    # Usage: nls_util.py input-file output-file language-map-file
    if len(argv) < 4:
        sys.stderr.write("Usage: " + argv[0] + " input-file output-file language-map-file\n")
        return 1

    # Messages pool.
    messages = {}

    input_file_path = argv[1]
    output_file_path = argv[2]
    langmap_file_path = argv[3]

    # Prepare for replace.
    messages = load_file(langmap_file_path, MAIN_DOMAIN)
    msgs = list_messages(messages, MAIN_DOMAIN)
    data = readall(input_file_path)
    replaced_count = 0

    # Replace key to message.
    for msg in msgs:
        data = data.replace(msg, get_message(messages, msg, MAIN_DOMAIN, msg))
        replaced_count += 1

    # Write data.
    writeall(output_file_path, data)

    sys.stderr.write(
        output_file_path + ": Successfully replaced " + str(replaced_count) + " words.\n"
        )
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
